### PR TITLE
Store: Update PriceInput component so that users can reset the value to an `initialValue`

### LINKED ISSUE #18961: Store components: Add a way for users to reset to the initial value on `PriceInput`
@jameskoster made a suggestion in another PR https://github.com/Automattic/wp-calypso/pull/18729#issuecomment-335857895 — On the refunding modal, we pre-fill the PriceInputs with the shipping cost, fee values, etc. But we don't otherwise show those values when editing, so:

> One concern is that if you edit the refund value for a fee (or shipping for that matter) but then want to revert back to the full amount, it's possible you'd forget what the original value was. Which would mean a refresh + loss of changes.

The suggestion was a "Max" button next to the input:

68747470733a2f2f636c6475702e636f6d2f4c3068384d324630637a2d3330303078333030302e706e67

### PR DESCRIPTION
**Pull request by ryelle**
Fixes #18961. This PR adds a button as a "suffix" to the `FormCurrencyInput` used in `PriceInput`, which when clicked resets the value of the `PriceInput` to a supplied prop `initialValue`. This also introduces a new function `getOrderFeeCost` (+ tests) to get the value for a single fee.

On the edit page, this is used on the shipping total field:
edit

On the refund screen, this is used on the shipping total, and each fee field:
refund

⚠️  _Note:_ This does mess with the alignment of the total refund value, but I'll be changing that to the new design in #18672.

**To Test: Edit**

- Edit an existing order, with or without taxes, with or without a shipping cost
- Change the shipping value
- Click the "undo" button to reset the shipping value
- It should reset to the shipping total (not including taxes, as taxes are handled on save)

**To Test: Refunds**

- Open an order, click Submit Refund
- If the order has fees, the fee total, value + taxes, should be pre-populated
- Change the fee value, then click the "undo" button
- Verify the value is fee total _with_ tax, and that the total refund value also updated to reflect the reset fee value
- Change the shipping total, then click the "undo" button
- Verify that the shipping total is reset to the value _with_ tax

### CODE REVIEW

File: client/extensions/woocommerce/app/order/order-payment/table.js [hunk 1]
```diff
@@ -257,6 +263,7 @@ class OrderRefundTable extends Component {
 						isEditable
 						currency={ order.currency }
 						label={ translate( 'Shipping' ) }
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/order/order-payment/table.js]: Hi! I've found a possible matching string that has already been translated **20** times:
`translate( 'Shipping address' )` *ES Score: 9*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/store-price-input-max-button)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/components/price-input/index.js [hunk 2]
```diff
@@ -21,31 +23,74 @@ import QuerySettingsGeneral from 'woocommerce/components/query-settings-general'
...
+			);
+		}
 		return (
 			<div>
 				<QuerySettingsGeneral siteId={ siteId } />
 				{ currencyObject ? (
 					<FormCurrencyInput
 						currencySymbolPrefix={ currencyObject.symbol }
-						value={ value }
+						currencySymbolSuffix={ resetButton }
```
Comment (timmyc) [inline client/extensions/woocommerce/components/price-input/index.js]: Neat - very creative usage here.

File: client/extensions/woocommerce/components/price-input/index.js [hunk 3]
```diff
@@ -21,31 +23,74 @@ import QuerySettingsGeneral from 'woocommerce/components/query-settings-general'
...
+						value={ this.state.value }
 						{ ...props }
 					/>
 				) : (
-					<FormTextInput value={ value } { ...omit( props, [ 'noWrap', 'min' ] ) } />
+					<FormTextInput value={ this.state.value } { ...omit( props, [ 'noWrap', 'min' ] ) } />
```
Comment (timmyc) [inline client/extensions/woocommerce/components/price-input/index.js]: Minor but `this.state.value` is used twice, maybe put it in a `const`?

**Comment by kellychoffman** [comment]
> The other item was do you think we should add a tooltip

I was just going to say this as well. I do wonder whether Jay's design with the word written out would be more easily understood and decrease the visual heaviness Timmy mentioned. It would also read more as a button with the blue link. Right now I think its too visually similar to the currency prefix, which acts only as a label. 

screen shot 2017-12-19 at 2 08 21 pm

**Comment by ryelle** [comment]
>  I do wonder whether Jay's design with the word written out would be more easily understood and decrease the visual heaviness

The button would still need to be this wide, so the alignment thing @timmyc mentioned (where it overlaps into the tax columns) is still going to be an issue. That reset link also looks pretty small, and would need to be a different color -- that blue is too light contrast for the white BG (2.6).

**Comment by kellychoffman** [comment]
>  that blue is too light contrast for the white BG (2.6).

Right, that makes sense. The largeness of the input doesn't bother me as much as the button that looks like a label. 

>  that blue is too light contrast for the white BG (2.6).

Oops, my bad I think I used blue-medium in that quick mock instead of the standard link color. 😬

**Comment by ryelle** [comment]
> the standard link color

Hm.... looks like the standard link color is also not high-enough contrast, esp for a small text - I'm assuming you mean `#3a86b9`, which is what the "Edit" link next to Billing Details is. What if it uses the grey on a white background? Looks less like a label (but IMO still not totally like an action)

screen shot 2017-12-20 at 2 15 07 pm

**Comment by allendav** [comment]
I'm noticing values that end in zero are formatted poorly, e.g. the $7.20 in shipping on this order:

screen shot 2017-12-20 at 1 06 03 pm

File: client/extensions/woocommerce/components/price-input/index.js [hunk 4]
```diff
@@ -21,31 +23,73 @@ import QuerySettingsGeneral from 'woocommerce/components/query-settings-general'
...
+	}
+
+	componentWillReceiveProps( { value } ) {
+		this.setState( { value } );
+	}
+
+	resetToInitial = () => {
+		const { initialValue, name = '' } = this.props;
+		this.setState( { value: initialValue }, () => {
+			// All uses of onChange expect and event object of this shape
```
Comment (allendav) [inline client/extensions/woocommerce/components/price-input/index.js]: and -> an ?

File: client/extensions/woocommerce/components/price-input/index.js [hunk 5]
```diff
@@ -21,31 +23,73 @@ import QuerySettingsGeneral from 'woocommerce/components/query-settings-general'
...
+			<div className="price-input">
 				<QuerySettingsGeneral siteId={ siteId } />
 				{ currencyObject ? (
 					<FormCurrencyInput
 						currencySymbolPrefix={ currencyObject.symbol }
-						value={ value }
-						{ ...props }
+						currencySymbolSuffix={ resetButton }
+						value={ this.state.value }
+						{ ...omit( props, [ 'dispatch', 'moment', 'numberFormat' ] ) }
```
Comment (allendav) [inline client/extensions/woocommerce/components/price-input/index.js]: Would it be appropriate to omit `translate` too?

Comment (ryelle) [reply client/extensions/woocommerce/components/price-input/index.js]: No need, it's already omitted by the way this.props is destructured above.

**Comment by jameskoster** [comment]
If we go with the text link it needs to be the regular blue used for all other links in Calypso. It's not WCAG AA friendly but that's separate issue. I agree with Kelly, there's not enough contrast with the prefix otherwise and it doesn't look like an interactive element.

I do think it's worth trying the `gridicons-refresh` icon though, in the same style as the [password view/hide input](https://wpcalypso.wordpress.com/devdocs/design/form-fields#password). I think the refresh icon would work better as this is resetting rather than undoing. To me, `undo` suggests some kind of multi-layered chronology which isn't really accurate here.

**Comment by ryelle** [comment]
@jameskoster I tried the refresh icon:

screen shot 2018-01-03 at 2 15 37 pm

**Comment by jameskoster** [comment]
Styled it up a bit. Couple of minor points;

* Can we only reveal the icon when the value has changed?
* Notice a weird issue (the 0 being added): https://cloudup.com/cBQAo-AMbOW

**Comment by ryelle** [comment]
> Can we only reveal the icon when the value has changed?

How's this?

reset-button

> the 0 being added

I was able to fix this in the refund modal, but the way the price is calculated in the order edit screen means it will always default to zero.

**Comment by jameskoster** [comment]
Looks good!

**Pull request merged**